feat(android-core): Initialize Android Core SDK (#1)
asgardeo/mobile-ui-sdks: android/README.md

3. In the created mobile application, go to the "Advanced" section and [enable the application native authentication for your Android application](https://is.docs.wso2.com/en/latest/guides/authentication/add-application-native-login/#enable-app-native-authentication).

### Installing the SDK

Add the latest released SDK in the `build.gradle` file of your Android application.

```groovy
dependencies {
    implementation 'io.asgardeo:android:1.0.0'
}
```

Add a redirect scheme in the Android application. You need to add the `appAuthRedirectScheme` in the application `build.gradle` file.

This should be consistent with the CallBack Url of the Service Provider that you configured in the Asgardeo.

For example, if you have configured the `callBackUrl` as `wso2sample://oauth2`, then the `appAuthRedirectScheme` should be `wso2sample`.

```groovy
android.defaultConfig.manifestPlaceholders = [
    'appAuthRedirectScheme': 'wso2sample'
]

> ```

Reviewer: Shall we make this a numbered list to clearly outline the steps?

Author: Fixed with https://github.com/asgardeo/mobile-ui-sdks/pull/1/commits/fe2efb415d877fcf11b14155642113e751069e6e

3. After that you need to create a new service account in the google project. You can create a Service Account for yourself with the following steps.

    1. Go to `IAM & Admin` -> `Service Accounts`.
    2. Click Create Service Account.
    3. Fill the name and click create and continue.
    4. You need to grant your service account the roles of Service Account User and Service Usage Consumer.
    5. Click continue and then Done
    6. You can see the service account added without keys, click : Actions -> Manage Keys for the service account.
    7. Click Add key and Select JSON.
    8. Save the JSON in secure place (We need this for Android Attestation Credentials for application metadata)

4. After that, Update Application Advanced properties. The application you created requires 2 properties to perform android attestation.

    1. Android package name
    2. androidAttestationServiceCredentials
       The JSON secret of Service Account downloaded. Note that this attribute is defined as a JSON object hence use the JSON key as it is.

> 5. Now you can call the "Google Play Integrity API" from your application and pass the integrity token that you will get from the API result to the integrityToken value of the AuthenticationCoreConfig

Reviewer: Shall we move this info to Asgardeo docs and add a link to docs here?

Author: there is an ongoing effort for this by another team, to bring docs related to app native authentication to Asgardeo docs.

```kotlin
private val asgardeoAuth: AsgardeoAuth = AsgardeoAuth.getInstance(
    AuthenticationCoreConfig(

>         baseUrl = "https://localhost:9443",

Reviewer: Let's use asgardeo specific base URLs here.

Author: Fixed with https://github.com/asgardeo/mobile-ui-sdks/pull/1/commits/67ea36cc9fef864e59221f887574afe50546eea3

        redirectUri = "wso2sample://oauth2",
        clientId = "<client_id>",
        scope = "openid",
        integrityToken = "<integrity_token>"
    )
)
```

## Use authenticators with the SDK

The Asgardeo Auth SDK provides out-of-the-box support for some authenticators, which are accessible via the `AuthenticationProvider`. Each of the following functions will emit the aforementioned `AuthenticationStates`, except for the `AuthenticationState.Initial`.

Before utilizing these authenticators, you need to integrate them into your application's login flow. You can find more information about this in the following link: [link_to_documentation].

### Use any authentication mechanism

If you are using any other authentication mechanism like email OTP, you can use the `authenticate` function. For this, you need to pass the authenticator id or authenticator which can be retrieved from the `authenticationFlow` returned from the `Authentication.Unauthenticated` state.

This can be used in two ways:

#### Authenticator parameters are known

> If you are aware of the authenticator parameters required for the authenticator (which can be found in the following link), you can directly call this function to authenticate the user with this authenticator.

Reviewer: `which can be found in the following link`. There doesn't seem to be a link embedded here.

Author: this document is in development, hence removing this sentence.

Author: Fixed with https://github.com/asgardeo/mobile-ui-sdks/pull/1/commits/67ea36cc9fef864e59221f887574afe50546eea3

> # Developer Documentation - For Android SDK

Reviewer: ```suggestion
# Asgardeo Android Auth SDK
```

Author: Fixed with https://github.com/asgardeo/mobile-ui-sdks/pull/1/commits/5a909401570da5c711812c97884fdb49afc7c335

## Introduction

> The Asgardeo Auth Android SDK enables Android applications (written in Kotlin) to utilize OpenID Connect (OIDC) authentication with the WSO2 Identity Server/Asgardeo serving as the Consumer Identity and Access Management (CIAM) Provider through API-based authentication. This SDK assists you in creating custom login flows directly within the applications themselves, without relying on browser redirects, thereby prioritizing user experience.

Reviewer: Kotlin libs can be used from Java Android apps as well, right? If yes, do we need to say that this sdk is only for apps written in Kotlin?

```suggestion
The Asgardeo Auth Android SDK enables Android applications to utilize OpenID Connect (OIDC) authentication with the WSO2 Identity Server/Asgardeo serving as the Consumer Identity and Access Management (CIAM) Provider through API-based authentication. This SDK assists you in creating custom login flows directly within the applications themselves, without relying on browser redirects, thereby prioritizing user experience.
```

Author: Yes you are correct, but using Kotlin libs in Java projects is bit cumbersome effort and not recommend by many, hence its better to mention the technology we have used for the developers information.

> The Asgardeo Auth Android SDK enables Android applications (written in Kotlin) to utilize OpenID Connect (OIDC) authentication with the WSO2 Identity Server/Asgardeo serving as the Consumer Identity and Access Management (CIAM) Provider through API-based authentication. This SDK assists you in creating custom login flows directly within the applications themselves, without relying on browser redirects, thereby prioritizing user experience.

Reviewer: Let's use the term  `app-native authentication`

```suggestion
The Asgardeo Auth Android SDK enables Android applications (written in Kotlin) to utilize OpenID Connect (OIDC) authentication with the WSO2 Identity Server/Asgardeo serving as the Consumer Identity and Access Management (CIAM) Provider through application-native authentication. This SDK assists you in creating custom login flows directly within the applications themselves, without relying on browser redirects, thereby prioritizing user experience.
```

## Prerequisite

- Supports Android applications written in Kotlin programming language.
- The minimum supported SDK is API level 26, compiled to API level 34. However, there are certain limitations when using specific authentication methods:

>     - Passkeys are supported on API level 34 and above.

Reviewer: ```suggestion
    - Passkeys are only supported on API level 34 and above.
```

    - Google authentication using the Credential Manager API is supported on API level 34 and above.

## Getting Started

> ### Register your application in /WSO2 Identity Server/Asgardeo

Reviewer: ```suggestion
### Register your application in Asgardeo
```

#### WSO2 Identity Server

1. Download the latest version of WSO2 Identity Server, and start the WSO2 Identity Server.
2. Register a Mobile Application to integrate your application with WSO2 Identity Server. You will obtain a `client_ID` from WSO2 Identity Server for your application which will need to be embedded later for the SDK integration. Also note the redirect URI that you used to create the application, this is also required for the SDK integration.
3. In the created mobile application go to the "Advanced" section and enable the application native authentication for your Android application.

#### Asgardeo

1. Register to Asgardeo and create an organization if you don't already have one. The organization name you choose will be referred to as `<org_name>` throughout this document.

> 2. Register a Mobile Application in Asgardeo to integrate your application with Asgardeo. You will obtain a `client_ID` from Asgardeo for your application which will need to be embedded later for the SDK integration. Also note the redirect URI that you used to create the application, this is also required for the SDK integration.

Reviewer: Maybe we can link to the relevant asgardeo docs as well?

Author: Fixed with https://github.com/asgardeo/mobile-ui-sdks/pull/1/commits/5a909401570da5c711812c97884fdb49afc7c335

> 3. In the created mobile application go to the "Advanced" section and enable the application native authentication for your Android application.

Reviewer: ```suggestion
3. In the created mobile application, go to the "Advanced" section and enable the application native authentication for your Android application.
```

### Installing the SDK

Add the latest released SDK in the `build.gradle` file of your Android application.

```groovy
dependencies {
    implementation 'io.asgardeo.android:1.0.0'
}
```

Add a redirect scheme in the Android application. You need to add the `appAuthRedirectScheme` in the application `build.gradle` file.

This should be consistent with the CallBack Url of the Service Provider that you configured in the WSO2 Identity Server/Asgardeo.

For example, if you have configured the `callBackUrl` as `wso2sample://oauth2`, then the `appAuthRedirectScheme` should be `wso2sample`.

```groovy
android.defaultConfig.manifestPlaceholders = [
    'appAuthRedirectScheme': 'wso2sample'
]
```

### Start the authentication process

1. First, you need to initialize the SDK object, `AsgardeoAuth`, to authenticate users into your application. This can be done in a repository if you are using an MVVM pattern in your application.

```kotlin
private val asgardeoAuth: AsgardeoAuth = AsgardeoAuth.getInstance(
    AuthenticationCoreConfig(

>         discoveryEndpoint = "https://localhost:9443//oauth2/token/.well-known/openid-configuration",

Reviewer: Will devs need to specify the discovery endpoint by themselves, when initializing sdk? I doubt that would be good DX. WDYT?

```suggestion
        discoveryEndpoint = "https://localhost:9443/oauth2/token/.well-known/openid-configuration",
```

        redirectUri = "wso2sample://oauth2",
        clientId = "<client_id>",

>         scope = "openid"

Reviewer: Shall we add an example of how to pass multiple scopes?

    )
)
```

> [!IMPORTANT]

> > Also if you are using an emulator to test out your application with a locally hosted IS instance, make sure to replace localhost with 10.0.2.2.

Reviewer: In accordance with our product strategy, let's remove the onprem instructions from asgardeo specific SDK repositories.

cc: @DonOmalVindula

Author: Fixed with https://github.com/asgardeo/mobile-ui-sdks/pull/1/commits/4cc9a7daa3b75815a614c19947544921e17b011d

>
> Also in the above scenario having only the discovery endpoint will not work, you need to have other endpoints setup as well, because from the discovery endpoint it will sent other endpoints with “localhost” which will not work with a locally hosted instance of IS.

> 

Reviewer: ```suggestion
```

val scope:String? = tokenProvider.getScope(context)
```

### Perform action based on the tokens

If you want to perform any action based on the tokens that are returned, you can use the `performAction` function in the `TokenProvider`.

```kotlin
tokenProvider.performAction(context) { accessToken, idToken, ->
    action(accessToken, idToken)
}
```

### Logout

If you want to perform a logout, you can call the `logout` function in the `AuthenticationProvider`. This will emit the state `AuthenticationState.Initial` if the logout is successful, and if an error occurs, it will emit `AuthenticationState.Error`.

```kotlin
authenticationProvider.logout(context)
```

### Client attestation
You can also use client attestation with the SDK as well.

#### How to setup client attestation for your application
1. In the created mobile application in the WSO2 Identity Server/Asgardeo, go to the advanced section and enable the client attestation.
2. You need to subscribe to the "Google Play Integrity API" from your Google cloud project, you can do this from Enabled API's & Services in your Google cloud project.

> 3. After that you need to create a new service account in the google project. You can create a Service Account for yourself with the following scopes.

Reviewer: ```suggestion
3. After that you need to create a new service account in the google project. You can create a Service Account for yourself with the following steps.
```

>     1. Go to IAM& Admin -> Service Accounts

Reviewer: ```suggestion
    1. Go to `IAM & Admin` -> `Service Accounts`.
```

```kotlin
class YourActivity : AppCompatActivity() {
    // ...

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        authenticationProvider.handleGoogleNativeLegacyAuthenticateResult(
            context,
            result.resultCode,
            result.data
        )
    }
    // ...
}
```

This will invoke the Google one tap authentication mechanism.

### Use Passkey authentication (Supports API 34 and above)

```kotlin
authenticationProvider.authenticateWithPasskey(
    context,
    authenticatorId = authenticator.authenticatorId,
)
```

### Use redirect based authentication

To perform redirect based authentication using a federated authenticator, first, you need to add the following code snippet to your application build.gradle file. This will add a separate activity that will handle the redirection on your behalf.

```gradle
android.defaultConfig.manifestPlaceholders = [
     ...
     'redirectUriHost': '<host>',
     'redirectUriScheme': '<scheme>'
     ...	
]
```

> After authenticating with the federated IdP, normally, the IdP will redirect the user to the WSO2 Identity Server/Asgardeo's common auth endpoint to continue the flow. However, with API-based authentication, this is changed. The IdP should redirect to the application. To support this, you should configure the deep link in the federated IdP side. Add that deep link in the `<data>` section. For example, if you are using the `wso2sample://oauth2` deep link, you should fill the `<data>` section as follows:

Reviewer: ```suggestion
After authenticating with the federated IdP, normally, the IdP will redirect the user to Asgardeo commonauth endpoint to continue the flow. However, with application-native authentication, this is changed. The IdP should redirect to the application. To support this, you should configure the deep link in the federated IdP side. Add that deep link in the `<data>` section. For example, if you are using the `wso2sample://oauth2` deep link, you should fill the `<data>` section as follows:
```
